core/domain/question_services.py

```diff
@@ -125,22 +125,28 @@ def get_questions_and_skill_descriptions_by_skill_ids(
             questions are to be returned. This value should be urlsafe.
 
     Returns:
-        list(Question), list(str), str. The list of questions and the
+        list(Question), list(list(str)), str. The list of questions and the
```

Comment:
To be clearer: I think we should get rid of the get_questions_and_skill_descriptions_by_skill_ids() function, and replace it with a get_questions_by_skill_ids() function, and let the caller handle retrieval of the skill descriptions separately.  Would that be possible?

Reply:
We're going to get rid of the get_questions_and_skill_descriptions_by_skill_ids() function because it's only used by pretest, and pretest should be integrated with the question player later.

For get_question_summaries_and_skill_descriptions() function, which has the same structure, I can't think of any better way. I think this function was retrieving them together because question_skill_link has the information of both question_id and skill_id, which makes it easier to retrieve both at the same time. If we get skill descriptions outside of this function, we will have to (a) fetch question_skill_links again, or  (b) get question from question_summaries.id, ,
 get linked_skill_ids from questions, and get skill descriptions from linked_skill_ids. Both ways are not efficient.

Reply:
Have the caller retrieve question summaries through get_question_summaries(). The caller then extracts the skill IDs from the question summaries, gets the corresponding skill descriptions, and returns a map of skill IDs to skill descriptions. Assemble a dict with both of these and send it to the frontend. The frontend can take care of the join (or alternatively just maintain its own skill-ID-to-description mapping).

(This assumes skill IDs are in the question summaries -- that's correct, right? If not, they probably should be.)

Reply:
I don't think linked_skill_ids is in questionSummary yet. It's only in Question. @vinitamurthi Could you confirm this?

Reply:
Skill IDs are not in question summaries..they are in question only

Reply:
@seanlip The only caller of the function get_questions_and_skill_descriptions_by_skill_ids() is the pretest backend handler. The pretest is going to completely move to use the question player directive. So why are we complicating this function and making further fetches? This has already been solved for in the question player

Reply:
@seanlip The question summary model only has the html content of the question and the question model has linked skill ids / state data and other information. So with that being the case it doesn't make sense to do multiple fetches right?

Reply:
Can we get them into question summaries?

Reply:
The question summaries are created to hold the question content data..all other data related to a question is kept in the question object, it doesn't really make sense to keep it in question summaries.

Reply:
I think the whole reason we created a separate QuestionSkillLink class was not to have skill ids in question summaries and each  link can be treated separately. If we add skill ids to question summaries, then there is no need of a question skill link class at all.
I forgot what the exact reasoning behind this was, but I think it was something along the lines of - questions are the most basic structure and so in the future multiple questions can be linked and deleted from a skill simultaneously in which case if they are in a single object, it could lead to race conditions?
Again, I am not sure of the exact reasoning, but we had decided not to have skill ids in questions or question summaries and only have their link as a separate class.

Reply:
Thanks @aks681!  I'm not sure I follow the reasoning outlined there, because the changes to the question's linked skill IDs would be batched and can be applied in one operation. I actually thought we created a QuestionSkillLink class to ease querying, and it seems to me that that class still has its uses -- it can additionally be used to store things like difficulty of the question in a fairly natural way.

That said, I'm getting the sense that this needs a bit more discussion, so @sophiewu6, feel free to go with any approach that works for this PR (for this issue) and we can address this in a follow-up once we figure out what the correct backend structure actually is.

Reply:
Sg, thanks!
Yeah, I think the reason we opted for link classes was because of ease of query.